Enable experimental usage of Microsoft.ReactNative.Managed as a nuget package (#5283)
microsoft/react-native-windows: docs/testing-nuget-packages.md

> ## Testing nuget package for new react-native init templates.

Reviewer: There's been back and forward discussion about the right thing to do, but some "How to" style docs have also been ending up in the wiki.

Author: @asklar has been guiding me to put docs here.
i.e. in the repo if they are for developers working in this repo *on* react-native-windows instead of wiki (whcih I beleive is also for that, and we should probably consolidate.
and the website and docs repo contains for developers *using* react-native-windows

Reviewer 2: @NickGerleman  we had this conversation about a month ago and you and Andrew pushed for putting the docs in the repo. We should pick one option and stick with it.

Reviewer: For me, in the physical repo or the wiki of the repo doesn't make too much of a difference (since both are repo local), but yeah, we should have that discussion. I know @kmelmon has been pushing some new stuff to the wiki, but wikis vs docs folder came up this morning again. My own preference is probably wiki over the docs folder but it's not a super strong one.

Reviewer: The other interesting case is stuff that's package local. E.g. README's need to go with a package. That seems potentially additive to other docs though.

> Todays default is to use the C++ and C# code from the react-native-windows npm package as source. We have been working on moving to a binary distribution model for this code via nuget. 

Reviewer 2: ```suggestion
Today's default is to use the C++ and C# code from the `react-native-windows` npm package as source. We have been working on moving to a binary distribution model for this code via NuGet. 
```

Today we support the following projects as a nuget package:
 * Microsoft.ReactNative
 * Microsoft.ReactNative.Managed

 Testing nuget packages is frequently a painful process as msbuild and nuget lack referential transparency. The steps in this document hope to alleviate some of the challenges.

## Complicating factors
* Nuget packages can only be loaded from a feed
  > These scripts help maintain a test feed on your local disk

> * Each change reuquires a version bump.

Reviewer 2: ```suggestion
* Each change requires a version bump.
```

>   Nuget caches by directory and simply checks if a directory with that version exists, regardless if the package has changed. It has implicit assumptions all up it's stack that versions are immutable. Therefore each time you create a new package, you are forced to bupm the version

Reviewer 2: ```suggestion
  NuGet caches by directory and simply checks if a directory with that version exists, regardless of whether the package has changed. It has implicit assumptions all up its stack that versions are immutable. Therefore each time you create a new package, you are forced to bump the version
```

   1. push vnext
   1. yarn link
      > if it fails with already being registred for anotehr enlistment you can run `yarn unlink`
1. Generate a test project
   1. open terminal in a test folder
   1. `npx react-native init MyProj`
   1. cd `MyProj`
   1. Apply the react-native-windows template:
      1. `node z:\src\r3\packages\react-native-windows-init\lib-commonjs\Cli.js --version 0.0.0 --useDevMode --overwrite  --language cs --experimentalNugetDependency --nugetTestFeed c:\temp\RnWNugetTesting\feed --nugetTestVersion 0.0.1-MyTest001 `
         > See below for a breakdown of the arguments:
1. Do your testing

## Changes:
* Changes to the binaries in the nuget package
  1. Rerun the builds for the slices you use
  1. Rerun publish nuget script bumping the version number by one
  1. Reapply the windows template with the new version number
  1. Do your tests
* Change to the template generation
  1. yarn build
  1. Reapply the windows template
  1. Do your tests

## Scripts
### PublishNugetPackagesLocally
This script takes more options for more customization: `<version> [<targetDir>] [<slices>] [<baseConfiguration>] [<basePlatform>]`.
      
Let's break down all arguments:

> * `targetDir`: Where the pacakges and feed should be placed. Default is `c:\temp\RnWNugetTesting`

Reviewer 2: ```suggestion
* `targetDir`: Where the packages and feed should be placed. Default is `c:\temp\RnWNugetTesting`
```

* `slices`: This is the set of slices to produce in powershell array syntax. The default is: `@('x64.Debug')`. An example of more slices would be: `@('x64.Debug', 'X86.Release', 'ARM64.Release')`
* `baseConfiguration`: The slice that is used for the non configuration specific build outputs. Default is: `Debug`
* `basePlatform`: The slice that is used for the non platform specific build outputs. Default is `x64`

### react-native-windows-init ... cli.js
This script has many options, but let's break down the ones from the sample:

> * `--version 0.0.0 --useDevMode`: The version doesn't matter when using dev mode, which means that the npm package is used that is registred as the dev package using `yarn link` per instructions above. This saves copying all the output files to your package directory, to your yarn cache. It is also a lot faser and since it is a link, you can make changes to your enlistment i.e. to `vnext\StyleSheets` and you don't have to reinstall between testing

Reviewer 3: >faser [](start = 306, length = 5)

nit: spelling

* `--overwrite`: This just overwites files that already exist instead of prompting. This is very usefull when running the script multiple times to test the templates or to code that emits the template.

> * `--language cs`: this is one of the public flags to geernate a cs project instead of a cpp project. 

Reviewer 3: >geernate [](start = 54, length = 8)

nit: spelling

> * `--experimentalNugetDependency`: This is the critical one to let the template emit referneces to nuget package rather than source pacakges. When we make it the default this will likely be renamed back to source for those that need a tight loop between test projects from template and . If it proves that the sampele and test apps we have int he repo satisfy this need, it can beremoved and the tempaltes can be clened up by removing a lto fo the conditionals bewteen the two.

Reviewer 3: >and  [](start = 282, length = 4)

nit: incomplete?

> * `--experimentalNugetDependency`: This is the critical one to let the template emit referneces to nuget package rather than source pacakges. When we make it the default this will likely be renamed back to source for those that need a tight loop between test projects from template and . If it proves that the sampele and test apps we have int he repo satisfy this need, it can beremoved and the tempaltes can be clened up by removing a lto fo the conditionals bewteen the two.

Reviewer 3: >sampele [](start = 310, length = 7)

nit: spelling

> * `--experimentalNugetDependency`: This is the critical one to let the template emit referneces to nuget package rather than source pacakges. When we make it the default this will likely be renamed back to source for those that need a tight loop between test projects from template and . If it proves that the sampele and test apps we have int he repo satisfy this need, it can beremoved and the tempaltes can be clened up by removing a lto fo the conditionals bewteen the two.

Reviewer 3: >int he [](start = 340, length = 6)

nit: spelling

> * `--experimentalNugetDependency`: This is the critical one to let the template emit referneces to nuget package rather than source pacakges. When we make it the default this will likely be renamed back to source for those that need a tight loop between test projects from template and . If it proves that the sampele and test apps we have int he repo satisfy this need, it can beremoved and the tempaltes can be clened up by removing a lto fo the conditionals bewteen the two.

Reviewer 3: >beremoved [](start = 378, length = 9)

nit: spelling

> * `--experimentalNugetDependency`: This is the critical one to let the template emit referneces to nuget package rather than source pacakges. When we make it the default this will likely be renamed back to source for those that need a tight loop between test projects from template and . If it proves that the sampele and test apps we have int he repo satisfy this need, it can beremoved and the tempaltes can be clened up by removing a lto fo the conditionals bewteen the two.

Reviewer 3: >clened [](start = 413, length = 6)

nit: spelling

> * `--experimentalNugetDependency`: This is the critical one to let the template emit referneces to nuget package rather than source pacakges. When we make it the default this will likely be renamed back to source for those that need a tight loop between test projects from template and . If it proves that the sampele and test apps we have int he repo satisfy this need, it can beremoved and the tempaltes can be clened up by removing a lto fo the conditionals bewteen the two.

Reviewer 3: >lto fo the conditionals bewteen [](start = 437, length = 31)

nit: spelling

* `--nugetTestFeed c:\temp\RnWNugetTesting\feed`: This flag adds an extra feed to the nuget.config so that for local testing you can use a different feed. This is the default location that is pushed to by the `PublishNugetPackagesLocally.cmd` script.

> * `--nugetTestVersion 0.0.1-MyTest001`: By default the nuget version matches the react-native version. For testing this can be cumbersome. Therefore this flag allows you to specify the version nubmer for the nuget packages. 

Reviewer 3: >nubmer [](start = 193, length = 6)

nit: spelling
